specification/0057-configuration-metadata-api.md

```diff
@@ -0,0 +1,230 @@
+unifies the configuration UI model among mods and allows mods to avoid reinventing the wheel. Exposing a common API to query config metadata
+allows config screens to be generated rather than relying on mods to manually construct screens using general-purpose GUI libraries.
+
+
+## Explanation
+
+### Definitions
+- **Setting**: A single value and its associated metadata.
+- **Category**: A logical grouping of Settings. A Setting may be a member of multiple Categories.
+- **Validation Group**: A logical grouping of Settings used for validation. A Setting may be a member of only one Validation Group.
+- **Value Converter**: An object that handles conversion between the value and UI state.
+- **Formatter**: A function that maps a value to Minecraft `Text` suitable for display.
+- **Setting Validator**: A function that is invoked to accept or reject the value of a single Setting.
+- **Group Validator**: A function that is invoked to accept or reject the set of values in a single Validation Group.
+- **Environment Policy**: A value that defines how a Setting interacts with the client and/or server.
+
+A new QSL module, `quilt_config_metadata`, contains the implementation of this API. The API is built on top of the Loader config API, specialized
+with metadata for Minecraft mods.
+
+### Settings API Detail
+
+A Setting has the following metadata.
+- Widget Type
+- Value Calculator
+- Formatter
+- Label
+- Tooltip
+- Categories
+- Group Validator
+- Environment Policy
+
+#### Widget Type
+
+```java
+public enum WidgetType {
+  TOGGLE,
+  CYCLE,
+  SELECT,
+  FREE_TEXT,
+  SLIDER,
+  NUMERIC,
+  UNORDERED_COLLECTION
+  ORDERED_COLLECTION;
+}
+
+MetadataType<WidgetType, ?> WIDGET_TYPE;
+```
+
+A Setting's Widget Type describes the type of widget that best represents the Setting.
+- `TOGGLE` widgets are binary, typically displayed as check boxes or buttons.
+- `CYCLE` widgets are small ordered lists of options, typically displayed as cycle buttons or radio buttons.
+- `SELECT` widgets are large unordered lists of options, typically displayed as a drop-down or searchable menu.
+- `FREE_TEXT` widgets are strings of text, typically displayed as a text box.
+- `SLIDER` widgets are imprecise numeric values, typically displayed as a slider.
+- `NUMERIC` widgets are precise numeric values, typically displayed as a text box.
+- `UNORDERED_COLLECTION` widgets are made up of a collection of values with no defined ordering, all with the same Widget Type.
+- `ORDERED_COLLECTION` widgets are made up of a sequence of values with a defined ordering, all with the same Widget Type.
+
+The most appropriate Widget Type for a given Setting depends on the Setting's underlying data type and on any constraints a Setting's
+value must obey.
+
+If a Setting's Widget Type is `UNORDERED_COLLECTION` or `ORDERED_COLLECTION`, additional data is stored `WIDGET_ELEMENT_TYPE`.
+
+```java
+MetadataType<WidgetType, ?> WIDGET_ELEMENT_TYPE;
+```
+
+#### Value Converter
```

Comment:
I also don't think this is necessary. `WIDGET_TYPE` should just be a `MetadataType<Function<TrackedValue<?>, Widget>>`, with different kinds of widgets having either constructors or helper methods for different parameter types. `ToggleWidget` for instance could have the following method:
```java
public static Widget create(TrackedValue<?> value) {
  Object defaultValue = value.value();

  if (defaultValue instanceof Boolean) {
    return new ToggleWidget(value.value());
  } else if (defaultValue instanceof Enum e && e.values().length == 2) {
    return new EnumToggleWidget(value.value());
  } else if (...) {
  
  } else {
    throw new WidgetCreationException(...); // Might be better to have some sort of result class instead of throwing errors, but ¯\_(ツ)_/¯ 
  }
}
```
The value would then simply call
```java
builder.metadata(WIDGET_TYPE, builder -> builder.factory(ToggleWidget::create))
```